# 01 - Loan Approval __ Data Audit

In [2]:
#importing libraries

import numpy as np
import pandas as pd

Loading my dataset

In [7]:
import os

print(os.getcwd())

/workspaces/loan-approval-prediction/notebooks


In [8]:
loan_data=pd.read_csv("/workspaces/loan-approval-prediction/data/raw/loan_data_new.csv")


# Viewing the structure of my data 

In [9]:
loan_data.head()

,Age,Gender,Education,Person Income,Employee Experience,Home Onwership,Loan Amount,Loan Intent,Loan interest Rate,Loan percentage,Credit History,Credit Score,Previous Loan,Loan Status
0,22,female,Master,71948,0,RENT,35000,PERSONAL,16.02,0.49,3,561,No,1
1,21,female,High School,12282,0,OWN,1000,EDUCATION,11.14,0.08,2,504,Yes,0
2,25,female,High School,12438,3,MORTGAGE,5500,MEDICAL,12.87,0.44,3,635,No,1
3,23,female,Bachelor,79753,0,RENT,35000,MEDICAL,15.23,0.44,2,675,No,1
4,24,male,Master,66135,1,RENT,35000,MEDICAL,14.27,0.53,4,586,No,1


In [11]:
loan_data.shape

(45000, 14)

In [12]:
loan_data.columns.tolist()

['Age',
 'Gender',
 'Education',
 'Person Income',
 'Employee Experience',
 'Home Onwership',
 'Loan Amount',
 'Loan Intent',
 'Loan interest Rate',
 'Loan percentage',
 'Credit History',
 'Credit Score',
 'Previous Loan',
 'Loan Status']

In [13]:
loan_data.info()

<class 'pandas.DataFrame'>
RangeIndex: 45000 entries, 0 to 44999
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   Age                  45000 non-null  int64  
 1   Gender               45000 non-null  str    
 2   Education            45000 non-null  str    
 3   Person Income        45000 non-null  int64  
 4   Employee Experience  45000 non-null  int64  
 5   Home Onwership       45000 non-null  str    
 6   Loan Amount          45000 non-null  int64  
 7   Loan Intent          45000 non-null  str    
 8   Loan interest Rate   45000 non-null  float64
 9   Loan percentage      45000 non-null  float64
 10  Credit History       45000 non-null  int64  
 11  Credit Score         45000 non-null  int64  
 12  Previous Loan        45000 non-null  str    
 13  Loan Status          45000 non-null  int64  
dtypes: float64(2), int64(7), str(5)
memory usage: 6.2 MB


In [14]:
loan_data.describe(include='all').T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
Age,45000.0,NaN,NaN,NaN,27.764178,6.045108,20.0,24.0,26.0,30.0,144.0
Gender,45000,2,male,24841,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Education,45000,5,Bachelor,13399,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Person Income,45000.0,NaN,NaN,NaN,80319.053222,80422.498632,8000.0,47204.0,67048.0,95789.25,7200766.0
Employee Experience,45000.0,NaN,NaN,NaN,5.410333,6.063532,0.0,1.0,4.0,8.0,125.0
Home Onwership,45000,4,RENT,23443,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Loan Amount,45000.0,NaN,NaN,NaN,9583.157556,6314.886691,500.0,5000.0,8000.0,12237.25,35000.0
Loan Intent,45000,6,EDUCATION,9153,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Loan interest Rate,45000.0,NaN,NaN,NaN,11.006606,2.978808,5.42,8.59,11.01,12.99,20.0
Loan percentage,45000.0,NaN,NaN,NaN,0.139725,0.087212,0.0,0.07,0.12,0.19,0.66


Insuring if there's no duplicated values

In [15]:
loan_data.duplicated().sum()

np.int64(0)

Insuring if there's no missing  values

In [16]:
missing_values = pd.DataFrame({
    "Missing Values": loan_data.isnull().sum(),
    "Percentage": loan_data.isna().mean()*100
})

missing_values.sort_values(by="Missing Values", ascending=False)
missing_values.sort_values("Percentage", ascending=False)

,Missing Values,Percentage
Age,0,0.0
Gender,0,0.0
Education,0,0.0
Person Income,0,0.0
Employee Experience,0,0.0
Home Onwership,0,0.0
Loan Amount,0,0.0
Loan Intent,0,0.0
Loan interest Rate,0,0.0
Loan percentage,0,0.0


# Full_Report of Data Audit 

Overview of the Dataset:

The loan approval dataset was examined as the first stage of the machine learning project to establish its structure, completeness, variable characteristics, and potential data-quality issues before undertaking exploratory analysis and predictive modelling.

The dataset contains 45,000 observations and 14 variables. Each observation represents a loan application, while the variables describe applicant demographic characteristics, educational background, income and employment information, loan characteristics, and credit-related information. The response variable is Loan Status, which represents the loan approval outcome.

The dataset therefore provides a suitable basis for investigating whether applicant, financial, loan, and credit characteristics can be used to predict loan approval status.

# Dataset Dimensions

The dimensions of the dataset were examined using:

loan_data.shape

The resulting dimensions were:

[
n = 45,000
]

observations and

[
p = 14
]

variables.

Thus, the dataset consists of 45,000 loan applications described by 14 variables.

The relatively large number of observations provides sufficient data for developing and comparing several classification algorithms while retaining observations for validation and testing.

# Variables in the Dataset

The dataset contains the following 14 variables:
Variable       |        Description  
Age            | Age of the loan      applicant
Education      | Highest reported education level 
Person Income  | Applicant's income
Employee Experience | Applicant's employment experience
Home Onwership | Applicant's home ownership status
Loan Amount    | Amount requested by the applicant
Loan Intent    | Purpose for which the loan is requested
Loan interest Rate  | Interest rate associated with the loan
Loan percentage     | Loan amount expressed relative to income
Credit History      | Length/history of the applicant's credit record
Credit Score   | Applicant's credit score
Previous Loan  | Whether the applicant has a previous loan
Loan Status    | Loan approval outcome and target variable 


Therefore variables can broadly be divided into four groups:
1. d target variable

The variables can broadly be divided into four groups:

3.1 Applicant characteristics

These include:

Age
Gender
Education
Employee Experience
Home Onwership

These variables describe the demographic, educational, employment, and housing characteristics of applicants.

2. Financial characteristics

The financial variables include:

Person Income
Loan Amount
Loan percentage
Loan interest Rate

These variables provide information about the applicant's financial position and the requested loan.

3. Credit characteristics

The credit-related variables are:

Credit History
Credit Score
Previous Loan

These variables capture aspects of the applicant's previous borrowing and creditworthiness.

4. Dependent/Target variable

Loan Status is the response variable to be predicted by the machine learning models.

# Data Types

The dataset contains:

7-integer variables,
2-floating-point variables, and
5-string/categorical variables.

The dataset contains no missing observations according to the info() output. All 45,000 observations are non-null for all 14 variables.

The categorical variables are:

Gender
Education
Home Onwership
Loan Intent
Previous Loan

The numerical variables are:

Age
Person Income
Employee Experience
Loan Amount
Loan interest Rate
Loan percentage
Credit History
Credit Score
Loan Status

Although Loan Status is stored as an integer, it should be treated as a binary categorical response variable, rather than as a continuous numerical variable.

# Missing-Value Assessment

Missing-value analysis was performed using both the number and percentage of missing observations.

The analysis showed that all 14 variables contain zero missing values. This means that missing-value imputation is not required for the current dataset. However, this finding does not imply that the dataset is free from data-quality problems. Implausible, extreme, or inconsistent values can exist even when all observations are complete.

Consequently, the subsequent data-quality assessment focuses on the validity and plausibility of the observed values.

# Duplicate-Record Assessment

The dataset was also examined for duplicate observations. Therefore, no completely duplicated records were identified.

This indicates that there are no exact duplicate rows requiring removal at this stage. Nevertheless, duplicate applicants cannot be completely ruled out because two applications could contain different values while belonging to the same individual. Such an assessment would require an applicant identifier, which is not included among the 14 variables.

# Descriptive Statistics

Descriptive statistics were examined for the numerical variables.

1 Age
The observed age has:
Mean =27.76 and SD=6.05.
The first quartile is 24 years, the median is 26 years, and the third quartile is 30 years.

However, the minimum and maximum values require particular attention. The minimum age is 20 years, while the maximum is 144 years. An age of 144 years is implausible for a loan applicant and strongly indicates the presence of an extreme or erroneous observation.

This is an important data-quality issue that should be investigated before modelling. The mean age of 27.76 years may also be affected by such extreme observations.

2. Person Income

Applicant income has a mean of approximately is 80,319.05 with a standard deviation of 80,422.50.
The median income is approximately 67,048, while the third quartile is approximately 95,789.

The most important finding is the maximum income is 7,200,766. This is substantially greater than the upper quartile and indicates a highly right-skewed distribution and/or the presence of extreme observations.

The difference between the mean and median also supports the possibility of substantial positive skewness.

The income variable therefore requires further investigation before modelling. Extreme values should not automatically be deleted; instead, their validity and potential influence should be assessed.

3. Employee Experience

Employee experience has a mean of approximately is 5.41 years and a median of 4 years. The minimum value is 0 years, which can be reasonable for applicants with no previous employment experience.

However, the maximum value is
125 years. This is highly implausible and indicates another potential data-quality problem.

Furthermore, employee experience should be considered in relation to applicant age. For example, an applicant aged 20 with 125 years of employment experience would be impossible.

Therefore, a consistency check between Age and Employee Experience should be performed.

4. Loan Amount

The requested loan amount has a mean of approximately is 9,583.16 and a standard deviation of approximately is 6,314.89.

The median loan amount is 8,000, while the first and third quartiles are 5,000 and 12,237.25, respectively.

The observed loan amounts range from  500 to 35,000. Unlike the age, income, and employment-experience variables, the loan amount does not immediately display an obviously impossible maximum. However, its distribution should still be examined graphically because the mean exceeds the median, suggesting some positive skewness.

5. Loan Interest Rate

The loan interest rate has a mean of approximately: 11.01% with a standard deviation of approximately 2.98. The minimum observed rate is 5.42%, while the maximum is 20%.

These values appear plausible within the context of the dataset, although the meaning and generation of this variable require further investigation.

In particular, it must be established whether the interest rate was available at the time the approval decision was made.

If the interest rate was assigned only after a loan was approved, using it as a predictor of approval could introduce data leakage. This issue must therefore be resolved before model development.

6. Loan Percentage

The Loan percentage variable has a mean of approximately: 0.1397 and a standard deviation of approximately: 0.0872. The median is 0.12 and the maximum is 0.66.

The first observations suggest that this variable may represent the ratio of loan amount to applicant income. For example, for the first observation:
\frac{35,000}/{71,948} = \approx0.4865, which is approximately the reported Loan percentage value of 0.49.

This suggests that Loan percentage may be a derived financial feature measuring the requested loan relative to applicant income.

This relationship should be formally verified before feature engineering is undertaken.

7. Credit History

Credit history has a mean of approximately:5.87 with a standard deviation of approximately: 3.88. The minimum is 2 and the maximum is 30. The median is 4, while the third quartile is 8.

The distribution should be examined further to determine whether the variable represents years of credit history or another measure of credit experience.

8. Credit Score

The credit score has a mean of approximately: 632.61 and a standard deviation of approximately: 50.44. The observed range is:
[
390 \leq \text{Credit Score} \leq 850.
]

The median credit score is 640. The observed range appears plausible for a credit-scoring variable and provides potentially useful predictive information for loan approval.

The relationship between credit score and loan status should be investigated during exploratory data analysis.

9. Categorical Variables

Five categorical variables are present.

Gender

There are two unique gender categories. The most frequent category is male, with 24,841 observations.

The complete frequency distribution should be examined to determine whether the categories are balanced.

Education

There are five education categories. Bachelor is the most frequent category, with 13,399 observations.

The remaining categories should be examined for frequency and consistency before encoding.

Home Ownership

There are four home-ownership categories, with RENT being the most frequent category at 23,443 observations.

Loan Intent

There are six loan-purpose categories, with EDUCATION being the most frequent category at 9,153 observations.

Previous Loan

There are two categories. Yes is the most frequent category, with 22,858 observations.

These categorical variables will require appropriate encoding before they can be supplied to most machine learning algorithms.

10. Target Variable: Loan Status

Loan Status is the proposed target variable for the classification problem.

The variable takes values between 0 and 1, with a mean of approximately: 0.2222. Because the variable is binary, its mean corresponds to the proportion of observations coded as 1.

Therefore, approximately 22.22% of observations belong to class 1, while approximately: 77.78% belong to class 0.

This indicates a substantial class imbalance.The exact interpretation of the classes must be established before proceeding. Specifically, it must be confirmed whether:
[
0= Rejected
1= Approved
]
or whether the coding is reversed.

This distinction is essential because it affects the interpretation of precision, recall, false-positive rates, false-negative rates, and the eventual lending decis

# Initial Data-Quality Findings

The audit has identified several issues requiring further investigation.

1. Implausible age

The maximum age is 144 years.

This is not a plausible applicant age and should be investigated.

2. Implausible employment experience

The maximum employee experience is 125 years.

This is also implausible and requires investigation.

3. Extreme income values

The maximum income is 7,200,766, which is substantially larger than the median and upper quartile.

The distribution should therefore be examined for extreme observations and skewness.

4. Possible derived variable

Loan percentage appears to be related to:

latex
[
\frac{\text{Loan Amount}}{\text{Person Income}}.
]

This should be formally verified.

5. Potential leakage

Loan interest Rate requires investigation to establish whether it was known at the point of loan approval.

6. Class imbalance

Only approximately 22.22% of observations belong to class 1.

Consequently, accuracy alone will not be sufficient for evaluating the machine learning models.

7. No missing observations

All variables contain complete observations.

8. No exact duplicate rows

No completely duplicated records were identified.


